front/lib/resources/membership_resource.ts

```diff
@@ -0,0 +1,370 @@
+import logger from "@app/logger/logger";
+
+type GetMembershipsOptions = RequireAtLeastOne<{
+  userIds: number[];
+  workspace: LightWorkspaceType;
+}> & {
+  roles?: MembershipRoleType[];
+  transaction?: Transaction;
+};
+
+// Attributes are marked as read-only to reflect the stateless nature of our Resource.
+// This design will be moved up to BaseResource once we transition away from Sequelize.
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+export interface MembershipResource
+  extends ReadonlyAttributesType<MembershipModel> {}
+export class MembershipResource extends BaseResource<MembershipModel> {
+  static model: ModelStatic<MembershipModel> = MembershipModel;
+
+  constructor(
+    model: ModelStatic<MembershipModel>,
+    blob: Attributes<MembershipModel>
+  ) {
+    super(MembershipModel, blob);
+  }
+
+  static async getActiveMemberships({
+    userIds,
+    workspace,
+    roles,
+    transaction,
+  }: GetMembershipsOptions): Promise<MembershipResource[]> {
+    const whereClause: WhereOptions<InferAttributes<MembershipModel>> = {
+      startAt: {
+        [Op.lte]: new Date(),
+      },
+      endAt: {
+        [Op.or]: [{ [Op.eq]: null }, { [Op.gte]: new Date() }],
+      },
+    };
+
+    if (userIds) {
+      whereClause.userId = userIds;
+    }
+    if (workspace) {
+      whereClause.workspaceId = workspace.id;
+    }
+    if (roles) {
+      whereClause.role = {
+        [Op.in]: roles,
+      };
+    }
+
+    const memberships = await MembershipModel.findAll({
+      where: whereClause,
+      transaction,
+    });
+
+    return memberships.map(
+      (membership) => new MembershipResource(MembershipModel, membership.get())
+    );
+  }
+
+  static async getLatestMemberships({
+    userIds,
+    workspace,
+    roles,
+    transaction,
+  }: GetMembershipsOptions): Promise<MembershipResource[]> {
+    const orderedResourcesFromModels = (resources: MembershipModel[]) =>
+      resources
+        .sort((a, b) => a.startAt.getTime() - b.startAt.getTime())
+        .map(
+          (resource) => new MembershipResource(MembershipModel, resource.get())
+        );
+
+    const where: WhereOptions<InferAttributes<MembershipModel>> = {};
+    if (roles) {
+      where.role = {
+        [Op.in]: roles,
+      };
+    }
+    if (userIds) {
+      where.userId = {
+        [Op.in]: userIds,
+      };
+    }
+    if (workspace) {
+      where.workspaceId = workspace.id;
+    }
+
+    if (!workspace && !userIds?.length) {
+      throw new Error("At least one of workspace or userIds must be provided.");
+    }
+    if (userIds && !userIds.length) return [];
+
+    // Get all the memberships matching the criteria.
+    const memberships = await MembershipModel.findAll({
+      where,
+      order: [["startAt", "DESC"]],
+      transaction,
+    });
+    // Then, we only keep the latest membership for each (user, workspace).
+    const latestMembershipByUserAndWorkspace = new Map<
```

Comment:
Your version is less performant with O(n^2) complexity and IMHO less readable

Reply:
On the perf side, I thought o(n^2) would not be a problem here, might have been a bit fast in that assesment I reckon :+1: 

On the readability, I'm genuinely surprised, this can make up for a good 80/20 debate :) :beer: 

I get that you don't like "reduce", but to the point of creating a custom 6-line function, then 11 lines to turn an array into a map, that you will then turn back into an array?

16 lines (omitting the map line that is common to both solutions), whereas a 1-line reduce that directly says "add to the array if it isn't already there"

Obv will not block approval for this of course, but glad to discuss that on friday :+1: